File: frontend/lib/solar-options.ts
export type SolarOptionKind = "individual_solar" | "vnm" | "gnm";

export type BillSolarPrefill = {
  analysis_id: string;
  connection_id: string;
  consumer_name: string | null;
  address: string | null;
  period_units_kwh: number;
  monthly_units: number;
  sanctioned_load_kw: number;
  current_monthly_bill_inr: number | null;
  tariff_code: string;
  discom: string;
  category: string;
  as_of: string;
  suggested_plant_kwp: number | null;
  bill_date: string | null;
  billing_period: string | null;
  billing_period_days?: number | null;
  billing_period_months?: number;
  is_multi_month_period?: boolean;
  period_consumption_note?: string | null;
};

export type SolarOptionCard = {
  option: SolarOptionKind;
  title: string;
  status: string;
  monthly_saving_inr: number | null;
  plant_kwp: number | null;
  message: string;
  official_next_step: string | null;
  missing_inputs: string[];
  warnings: string[];
  result: Record<string, unknown>;
  intelligence_report: SolarIntelligenceReport | null;
};

export type ReportMetric = {
  label: string;
  value: string;
  detail: string | null;
};

export type ReportSection = {
  id: string;
  title: string;
  metrics: ReportMetric[];
};

export type SolarIntelligenceReport = {
  option: SolarOptionKind;
  title: string;
  status: string;
  headline: string;
  location_line: string;
  property_type: string;
  sections: ReportSection[];
  disclaimer: string;
  actions: string[];
};

export type BillLineItem = {
  code: string;
  label: string;
  amount: number;
  detail: string | null;
};

export type BillScenario = {
  title: string;
  subtitle: string | null;
  lines: BillLineItem[];
  total: number;
  units_kwh: number | null;
  notes: string[];
};

export type MonthlyBillEstimate = {
  month_index: number;
  month_label: string;
  calendar_month: number;
  seasonal_factor: number;
  estimated_units_kwh: number;
  estimated_bescom_bill_inr: number;
  estimated_vnm_bill_inr: number;
  estimated_saving_inr: number;
};

export type VNMMethodology = {
  monthly_baseline_kwh: number;
  coverage_fraction: number;
  coverage_label: string;
  coverage_source: string;
  illustrative_plant_kwp: number | null;
  monthly_kwh_per_kwp: number | null;
  illustrative_rate_inr_per_kwh: number;
  gst_percent: number;
  seasonal_model_label: string;
  steps: string[];
};

export type VNMSetupCost = {
  label: string;
  amount_inr: number;
  detail: string;
  per_flat_inr: number | null;
};

export type VNMComparison = {
  provider: string;
  provider_website: string | null;
  sanctioned_load_kw: number;
  billing_period: string | null;
  period_units_kwh: number;
  monthly_units: number;
  billing_period_months?: number;
  is_multi_month_period?: boolean;
  period_consumption_note?: string | null;
  current_bill_total_inr: number;
  expected_vnm_solar_credit_kwh: number | null;
  needs_expected_credit: boolean;
  credit_input_prompt: string | null;
  scenario_label: string;
  solar_kwh_credited: number;
  residual_grid_kwh: number;
  estimated_generation_kwh: number;
  surplus_kwh: number;
  illustrative_coverage_fraction: number;
  coverage_source: string;
  illustrative_plant_kwp: number;
  monthly_kwh_per_kwp: number;
  plant_slider_min_kwp: number;
  plant_slider_max_kwp: number;
  plant_slider_step_kwp: number;
  default_plant_kwp: number;
  surplus_note: string | null;
  illustrative_rate_inr_per_kwh: number;
  gst_percent: number;
  vnm_energy_cost_inr: number;
  vnm_gst_inr: number;
  vnm_service_total_inr: number;
  residual_bescom_charges_inr: number;
  has_gruha_jyothi: boolean;
  gruha_jyothi_note: string | null;
  period_difference_inr: number;
  monthly_difference_inr: number;
  annual_difference_inr: number;
  is_vnm_cheaper: boolean;
  period_saving_inr: number;
  period_increase_inr: number;
  monthly_saving_inr: number;
  monthly_increase_inr: number;
  annual_saving_inr: number;
  annual_increase_inr: number;
  current_bill: BillScenario;
  vnm_bill: BillScenario;
  calculation_detail_lines: BillLineItem[];
  monthly_chart: MonthlyBillEstimate[];
  methodology: VNMMethodology | null;
  cta_primary: string;
  cta_secondary: string;
  cta_url: string;
  assumptions: string[];
  disclaimer: string;
};

export type SolarOptionsComparison = {
  analysis_id: string;
  prefill: BillSolarPrefill;
  options: SolarOptionCard[];
  best_option: SolarOptionKind | null;
  vnm_comparison: VNMComparison | null;
  disclaimer: string;
  message: string;
};

export type SolarOptionsResponse = {
  message: string;
  comparison: SolarOptionsComparison;
};

export type VNMParticipantInput = {
  connection_id: string;
  category?: string;
  monthly_units?: number;
  sanctioned_load_kw?: number;
  procurement_share_percent?: number;
};

export type GNMInstallationInput = {
  connection_id: string;
  category?: string;
  monthly_units?: number;
  sanctioned_load_kw?: number;
  priority?: number;
  is_host?: boolean;
};

export type CompareSolarOptionsPayload = {
  plant?: {
    proposed_kwp?: number;
    roof_area_m2?: number;
    same_discom_area?: boolean;
    same_consumer_name?: boolean;
  };
  vnm_participants?: VNMParticipantInput[];
  gnm_installations?: GNMInstallationInput[];
  include_individual_solar?: boolean;
  include_vnm?: boolean;
  include_gnm?: boolean;
  expected_vnm_solar_credit_kwh?: number;
  illustrative_plant_kwp?: number;
  illustrative_coverage_fraction?: number;
};

/** 1 sq ft ≈ 0.092903 m² */
export const SQFT_TO_M2 = 0.092903;

export const OPTION_LABELS: Record<SolarOptionKind, string> = {
  individual_solar: "Individual rooftop",
  vnm: "Virtual Net Metering (VNM)",
  gnm: "Group Net Metering (GNM)",
};

export function statusClass(status: string): string {
  if (status === "ESTIMATED" || status === "POTENTIALLY_SUITABLE") return "ok";
  if (status === "INSUFFICIENT_INFORMATION" || status === "NO_ROOF") return "warn";
  return "bad";
}

File: app/domain/models/solar_options.py
from __future__ import annotations

from datetime import date
from typing import Any, Literal

from pydantic import BaseModel, Field

from app.domain.models.gnm import GNMInstallationInput, GNMPlantInput
from app.domain.models.solar_intelligence_report import SolarIntelligenceReport
from app.domain.models.vnm_comparison import VNMComparisonView
from app.domain.models.vnm import VNMParticipantInput, VNMPlantInput


class SolarOptionsPlantInput(BaseModel):
    proposed_kwp: float | None = Field(
        default=None,
        ge=0,
        description="Plant size in kWp. If omitted, derived from bill consumption.",
    )
    roof_area_m2: float | None = Field(
        default=None,
        ge=0,
        description="Usable roof area for individual rooftop sizing.",
    )
    same_discom_area: bool = True
    same_consumer_name: bool = True
    estimated_monthly_generation_kwh: float | None = Field(default=None, ge=0)


class VNMParticipantOverride(BaseModel):
    connection_id: str
    category: str = "DOMESTIC"
    monthly_units: float | None = Field(default=None, ge=0)
    sanctioned_load_kw: float | None = Field(default=None, ge=0)
    procurement_share_percent: float | None = Field(default=None, ge=0, le=100)


class GNMInstallationOverride(BaseModel):
    connection_id: str
    category: str = "DOMESTIC"
    monthly_units: float | None = Field(default=None, ge=0)
    sanctioned_load_kw: float | None = Field(default=None, ge=0)
    priority: int | None = Field(default=None, ge=1)
    is_host: bool = False


class CompareSolarOptionsRequest(BaseModel):
    plant: SolarOptionsPlantInput = Field(default_factory=SolarOptionsPlantInput)
    vnm_participants: list[VNMParticipantOverride] = Field(default_factory=list)
    gnm_installations: list[GNMInstallationOverride] = Field(default_factory=list)
    include_individual_solar: bool = False
    include_vnm: bool = True
    include_gnm: bool = False
    expected_vnm_solar_credit_kwh: float | None = Field(
        default=None,
        ge=0,
        description=(
            "Expected VNM solar credit for the billing period (kWh), from the "
            "VNM provider/society — not derived from consumption or sanctioned load."
        ),
    )


class BillSolarPrefill(BaseModel):
    analysis_id: str
    connection_id: str
    consumer_name: str | None = None
    address: str | None = None
    # Raw kWh on the bill for the full billing period (may span >1 month).
    period_units_kwh: float
    # Average kWh/month when period spans multiple months; else same as period_units.
    monthly_units: float
    sanctioned_load_kw: float
    current_monthly_bill_inr: float | None = None
    tariff_code: str
    discom: str
    category: str
    as_of: date
    suggested_plant_kwp: float | None = None
    bill_date: str | None = None
    billing_period: str | None = None
    billing_period_days: int | None = None
    billing_period_months: float = 1.0
    is_multi_month_period: bool = False
    period_consumption_note: str | None = None


class SolarOptionCard(BaseModel):
    option: Literal["individual_solar", "vnm", "gnm"]
    title: str
    status: str
    monthly_saving_inr: float | None = None
    plant_kwp: float | None = None
    message: str
    official_next_step: str | None = None
    missing_inputs: list[str] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)
    result: dict[str, Any] = Field(default_factory=dict)
    intelligence_report: SolarIntelligenceReport | None = None


class SolarOptionsComparisonView(BaseModel):
    analysis_id: str
    prefill: BillSolarPrefill
    options: list[SolarOptionCard] = Field(default_factory=list)
    best_option: Literal["individual_solar", "vnm", "gnm"] | None = None
    vnm_comparison: VNMComparisonView | None = None
    disclaimer: str
    message: str


def build_assumed_vnm_participants(
    prefill: BillSolarPrefill,
    *,
    typical_flats: int = 20,
) -> list[VNMParticipantInput]:
    """Society model — user does not enter neighbour details."""
    flats = max(2, typical_flats)
    user_share = round(100.0 / flats, 4)
    other_share = round(100.0 - user_share, 4)
    participants = [
        VNMParticipantInput(
            connection_id=prefill.connection_id,
            category=prefill.category,
            sanctioned_load_kw=prefill.sanctioned_load_kw,
            monthly_units=prefill.monthly_units,
            procurement_share_percent=user_share,
        ),
        VNMParticipantInput(
            connection_id="Other flats (assumed)",
            category=prefill.category,
            sanctioned_load_kw=prefill.sanctioned_load_kw,
            monthly_units=prefill.monthly_units * (flats - 1),
            procurement_share_percent=other_share,
        ),
    ]
    return participants


def build_vnm_participants(
    prefill: BillSolarPrefill,
    extras: list[VNMParticipantOverride],
) -> list[VNMParticipantInput]:
    participants: list[VNMParticipantInput] = []
    bill_share = 100.0
    if extras:
        extra_shares = sum(p.procurement_share_percent or 0 for p in extras)
        bill_share = max(0.0, 100.0 - extra_shares)

    participants.append(
        VNMParticipantInput(
            connection_id=prefill.connection_id,
            category=prefill.category,
            sanctioned_load_kw=prefill.sanctioned_load_kw,
            monthly_units=prefill.monthly_units,
            procurement_share_percent=bill_share,
        )
    )
    for extra in extras:
        participants.append(
            VNMParticipantInput(
                connection_id=extra.connection_id,
                category=extra.category,
                sanctioned_load_kw=extra.sanctioned_load_kw or prefill.sanctioned_load_kw,
                monthly_units=extra.monthly_units or prefill.monthly_units,
                procurement_share_percent=extra.procurement_share_percent or 0.0,
            )
        )
    return participants


def build_gnm_installations(
    prefill: BillSolarPrefill,
    extras: list[GNMInstallationOverride],
) -> list[GNMInstallationInput]:
    installations: list[GNMInstallationInput] = []
    has_host = any(e.is_host for e in extras)
    installations.append(
        GNMInstallationInput(
            connection_id=prefill.connection_id,
            category=prefill.category,
            sanctioned_load_kw=prefill.sanctioned_load_kw,
            monthly_units=prefill.monthly_units,
            priority=1,
            is_host=not has_host,
        )
    )
    for i, extra in enumerate(extras, start=2):
        installations.append(
            GNMInstallationInput(
                connection_id=extra.connection_id,
                category=extra.category,
                sanctioned_load_kw=extra.sanctioned_load_kw or prefill.sanctioned_load_kw,
                monthly_units=extra.monthly_units or prefill.monthly_units,
                priority=extra.priority or i,
                is_host=extra.is_host,
            )
        )
    return installations


def build_vnm_plant(
    plant: SolarOptionsPlantInput,
    proposed_kwp: float,
) -> VNMPlantInput:
    return VNMPlantInput(
        proposed_kwp=proposed_kwp,
        same_discom_area=plant.same_discom_area,
        estimated_monthly_generation_kwh=plant.estimated_monthly_generation_kwh,
    )


def build_gnm_plant(
    plant: SolarOptionsPlantInput,
    proposed_kwp: float,
) -> GNMPlantInput:
    return GNMPlantInput(
        proposed_kwp=proposed_kwp,
        same_discom_area=plant.same_discom_area,
        same_consumer_name=plant.same_consumer_name,
        estimated_monthly_generation_kwh=plant.estimated_monthly_generation_kwh,
    )
